商品購入機能の実装

95yuki0305 opened the pull request:
# What
商品購入機能の実装

# Why
商品購入機能を使用できるようにするため

95yuki0305 committed:
Formオブジェクト作成

95yuki0305 committed:
購入機能画面のビューファイルを編集

95yuki0305 committed:
各テーブルに保存する

95yuki0305 committed:
各エラーメッセージの表示

95yuki0305 committed:
購入したあとの商品に関する表示の変更

商品が売れていればsold outを表示
商品が売れていれば、各条件に応じて編集・削除・購入画面に遷移できない

95yuki0305 committed:
クレジットカード決済導入

95yuki0305 committed:
ダミーデータ作成

95yuki0305 committed:
order_addressモデルのテストコード実装

95yuki0305 commented:
# What
- 正しく情報を記入すると、商品の購入ができる動画(https://gyazo.com/2dc39d41354413e2ca64cd2dd004a130)
- 入力に問題のある状態では商品の購入はできず、エラーメッセージが出力される動画
(https://gyazo.com/48385f6b47cfd10afe854e9460ab4ded)
- ログアウト状態のユーザーが、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画(https://gyazo.com/8348d139b9044b8f448bc4627b7ceecc)
- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
(https://gyazo.com/d6001bae2748b5bf58e74b648ac6d22e)
- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
(https://gyazo.com/43bac3cfd8f7503ae1eee6b1ae17d17b)
- 売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
(https://gyazo.com/2dc39d41354413e2ca64cd2dd004a130)
- 売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
(https://gyazo.com/a5546c5b0845d0fb58564c0a840e00a6)
- ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（商品詳細機能実装時に未実装であった場合）
(https://gyazo.com/88115e5df68a83cd6993c78d23fea810)
- ログイン状態の出品者以外のユーザーでも、売却済みの商品に対しては「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
(https://gyazo.com/a5546c5b0845d0fb58564c0a840e00a6)
- ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
(https://gyazo.com/88115e5df68a83cd6993c78d23fea810)
- テスト結果の画像(https://gyazo.com/ccca1d057e07a30a1e38ebcebae641e8)

# Why
商品購入機能の確認をしてもらうため

d-kazu4k commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def index
+    @order_address = OrderAddress.new
+    @product = Product.find(params[:product_id])
```
```ruby
@product = Product.find(params[:product_id])
```
上記ですが、同ファイル内に複数記述されています。
メソッド化して、`before_action`で呼び出せるようにリファクタリングしましょう！

95yuki0305 replied:
修正しました。

d-kazu4k commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,41 @@
+  def index
+    @order_address = OrderAddress.new
+    @product = Product.find(params[:product_id])
+    if current_user == @product.user
+      redirect_to root_path
+    elsif @product.order.present?
+      redirect_to root_path
+    end
```
```ruby
    if current_user == @product.user
      redirect_to root_path
    elsif @product.order.present?
      redirect_to root_path
    end
```
上記もメソッド化してcreateアクションでも処理されるようにしましょう！
また、`redirect_to root_path`が2回記述されているため、リファクタリングができそうです！
下記をご参考いただき、修正してみましょう！
```ruby
    if current_user == @product.user ★ @product.order.present?
      redirect_to root_path
    end
```
★に演算子を入れてリファクタリングしましょう！

95yuki0305 replied:
修正しました。

d-kazu4k commented on app/views/products/show.html.erb:
```diff
@@ -25,13 +27,14 @@
-        <p class="or-text">or</p>
-        <%= link_to "削除", product_path, method: :delete, class:"item-destroy" %>
-
+        <% if @product.order.blank? %>
+          <%= link_to "商品の編集", edit_product_path, method: :get, class: "item-red-btn" %>
+          <p class="or-text">or</p>
+          <%= link_to "削除", product_path, method: :delete, class:"item-destroy" %>
+        <% end %>
       <%# 商品が売れていない場合はこちらを表示しましょう %>
-      <% else %>
-        <%= link_to "購入画面に進む", "#" ,class:"item-red-btn"%>
+      <% elsif @product.order.blank? %>
+        <%= link_to "購入画面に進む", product_orders_path(@product.id), class:"item-red-btn" %>
       <% end %>
     <% end %>
```
```ruby
    <% if user_signed_in? %>
      <% if current_user.id == @product.user.id %>
        <% if @product.order.blank? %>
          <%= link_to "商品の編集", edit_product_path, method: :get, class: "item-red-btn" %>
          <p class="or-text">or</p>
          <%= link_to "削除", product_path, method: :delete, class:"item-destroy" %>
        <% end %>
      <%# 商品が売れていない場合はこちらを表示しましょう %>
      <% elsif @product.order.blank? %>
        <%= link_to "購入画面に進む", product_orders_path(@product.id), class:"item-red-btn" %>
      <% end %>
    <% end %>
```
上記ですが`@product.order.blank?`が2回記述されておりますので、リファクタリングしてみましょう！

満たしたいこととしては下記だと存じます。

- ログイン状態かつ売却されていない商品にはボタンを表示する
- 出品者であれば、「商品の編集」と「削除」ボタンが表示される
- 出品者でなければ、「購入画面に進む」ボタンが表示される

考え方としては、２つに分けられそうです。

1. ボタン表示の有無
2. 表示されるボタンの種類

そうした場合に、このような書き方ができそうです。
```ruby
<%# ボタン表示の有無 %>
  <% if user_signed_in? ★ @product.order.blank? %>
<%# 表示されるボタンの種類 %>
    <% if current_user.id == @item.user_id %>
      <%= link_to "商品の編集", edit_product_path, method: :get, class: "item-red-btn" %>
      <p class="or-text">or</p>
      <%= link_to "削除",product_path, method: :delete, class:"item-destroy" %>
    <% else %> 
      <%= link_to "購入画面に進む", product_orders_path(@product.id),class:"item-red-btn"%>
    <%end%>
  <%end%>
```
※★の部分には、演算子が入ります。条件を満たせるようにチャレンジしてみましょう！
※`@product.order.blank?`の記述は、記述方法によっては変更が必要な場合がございます。
上記ご参考にしていただき、修正をしてみましょう！

95yuki0305 replied:
修正しました。

d-kazu4k commented on spec/factories/order_address.rb:
```diff
@@ -0,0 +1,11 @@
+FactoryBot.define do
+  factory :order_address do
+    postal_code       {"123-4567"}
+    prefecture_id     {(2)}
```
```
prefecture_id {(2)}
```
上記の（）は外しましょう！

95yuki0305 replied:
修正しました。

d-kazu4k commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,82 @@
+        expect(@order_address).to be_valid
+      end
+    end
+    context '商品を購入できないとき' do
```
異常系テストに下記を追加しましょう！

- 電話番号が英数混合では登録できない
- user_idが空では登録できない
- user_idが空では登録できない

95yuki0305 replied:
修正しました。

d-kazu4k commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,82 @@
+  end
+
+  describe '商品購入' do
+    context '商品を購入できるとき' do
```
正常系テストに下記を追加しましょう！

- 建物名が空でも登録できる

95yuki0305 replied:
修正しました。

d-kazu4k reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

95yuki0305 committed:
メソッド化して、before_actionを設定

@product = Product.find(params[:product_id])

95yuki0305 committed:
メソッド化して、before_actionに設定(createアクションでも処理)

  def redirect_root
    if current_user == @product.user || @product.order.present?
      redirect_to root_path
    end
  end

95yuki0305 committed:
条件分岐式をリファクタリング

95yuki0305 committed:
不要な記述の削除

95yuki0305 committed:
異常系テストコードを追加

・phone_numberが英数混合では購入できない
・user_idが空では購入できない
・product_idが空では購入できない

95yuki0305 committed:
正常系テストコードを追加

・building_nameが空でも購入できる

chobi555 commented on app/controllers/products_controller.rb:
```diff
@@ -24,6 +24,9 @@ def show
   end
 
   def edit
+    unless @product.user == @product.order
+      redirect_to root_path
+    end
```
現在ですと、出品者と商品購入者が一致していたらという条件分岐になっています。
商品が売れていた場合の条件分岐はorders_controllerの条件分岐を参考に変更しておきましょう。

また、こちらの条件分岐はprivateメソッドのredirect_rootと論理演算子を用いてまとめることができます。
まとめてしまいましょう。

95yuki0305 replied:
修正しました。

chobi555 reviewed (commented):
レビュー依頼ありがとうございます！
コメント致しました。
確認後、修正をお願い致します。

95yuki0305 committed:
条件分岐を変更

shuki-yamamoto reviewed (commented):
大変お待たせ致しました。
レビュー依頼ありがとうございます。
修正点確認させていただきました。
LGTMとさせていただきます！

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
商品購入機能の実装お疲れさまでした！
